### ISSUE TITLE
Error when customizing the binding configuration for consumers with DLQ in 2021.0.3 Spring Cloud

### ISSUE THREAD

**Issue by ferblaca** [spring-cloud/spring-cloud-stream#2417]
We have detected an unexpected behaviour with binding consumers with DLQ configured for the Kafka binder when upgrading to Spring Cloud version `2021.0.3` from version `2021.0.2`.

In our applications we have configured `ConsumerConfigCustomizer` and `ProducerConfigCustomizer` to set a specific `client.id` for each Kafka consumer and producer with the `bindingName` parameter:

```
  @Bean
  public ConsumerConfigCustomizer consumerConfigCustomizer() {
    return (consumerProperties, bindingName, destination) -> {
      consumerProperties.put(ConsumerConfig.CLIENT_ID_CONFIG, bindingName);
    };
  }

  @Bean
  public ProducerConfigCustomizer producerConfigCustomizer() {
    return (producerProperties, bindingName, destination) -> {
      producerProperties.put(ProducerConfig.CLIENT_ID_CONFIG, bindingName);
    };
  }
```

The problem is that when upgrading to Spring Cloud version `2021.0.3`, when initializing the consumer with DLQ, apparently a producer is also configured (I guess to send the message to the DLT topic when exceeding retries against a consumer error), but the `bindingName` parameter now arrives null:

```
2022-05-31 18:33:12.178 ERROR 21433 --- [           main] o.s.cloud.stream.binding.BindingService  : Failed to create consumer binding; retrying in 30 seconds

org.springframework.cloud.stream.binder.BinderException: Exception thrown while starting consumer: 
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindConsumer(AbstractMessageChannelBinder.java:480) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindConsumer(AbstractMessageChannelBinder.java:92) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.AbstractBinder.bindConsumer(AbstractBinder.java:143) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binding.BindingService.doBindConsumer(BindingService.java:184) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binding.BindingService.bindConsumer(BindingService.java:137) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binding.AbstractBindableProxyFactory.createAndBindInputs(AbstractBindableProxyFactory.java:118) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binding.InputBindingLifecycle.doStartWithBindable(InputBindingLifecycle.java:58) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at java.base/java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:608) ~[na:na]
	at org.springframework.cloud.stream.binding.AbstractBindingLifecycle.start(AbstractBindingLifecycle.java:57) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binding.InputBindingLifecycle.start(InputBindingLifecycle.java:34) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:178) ~[spring-context-5.3.20.jar:5.3.20]
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:54) ~[spring-context-5.3.20.jar:5.3.20]
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:356) ~[spring-context-5.3.20.jar:5.3.20]
	at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[na:na]
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:155) ~[spring-context-5.3.20.jar:5.3.20]
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:123) ~[spring-context-5.3.20.jar:5.3.20]
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:935) ~[spring-context-5.3.20.jar:5.3.20]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:586) ~[spring-context-5.3.20.jar:5.3.20]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145) ~[spring-boot-2.6.8.jar:2.6.8]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:745) ~[spring-boot-2.6.8.jar:2.6.8]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:420) ~[spring-boot-2.6.8.jar:2.6.8]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:307) ~[spring-boot-2.6.8.jar:2.6.8]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1317) ~[spring-boot-2.6.8.jar:2.6.8]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306) ~[spring-boot-2.6.8.jar:2.6.8]
	at com.example.demoStreamKafka.DemoStreamKafkaApplication.main(DemoStreamKafkaApplication.java:10) ~[classes/:na]
Caused by: java.lang.NullPointerException: null
	at java.base/java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011) ~[na:na]
	at java.base/java.util.concurrent.ConcurrentHashMap.putAll(ConcurrentHashMap.java:1089) ~[na:na]
	at java.base/java.util.concurrent.ConcurrentHashMap.<init>(ConcurrentHashMap.java:852) ~[na:na]
	at org.springframework.kafka.core.DefaultKafkaProducerFactory.<init>(DefaultKafkaProducerFactory.java:205) ~[spring-kafka-2.8.6.jar:2.8.6]
	at org.springframework.kafka.core.DefaultKafkaProducerFactory.<init>(DefaultKafkaProducerFactory.java:168) ~[spring-kafka-2.8.6.jar:2.8.6]
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.getProducerFactory(KafkaMessageChannelBinder.java:574) ~[spring-cloud-stream-binder-kafka-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.getErrorMessageHandler(KafkaMessageChannelBinder.java:1181) ~[spring-cloud-stream-binder-kafka-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.getErrorMessageHandler(KafkaMessageChannelBinder.java:163) ~[spring-cloud-stream-binder-kafka-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.registerErrorInfrastructure(AbstractMessageChannelBinder.java:716) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.registerErrorInfrastructure(AbstractMessageChannelBinder.java:657) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createConsumerEndpoint(KafkaMessageChannelBinder.java:735) ~[spring-cloud-stream-binder-kafka-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.kafka.KafkaMessageChannelBinder.createConsumerEndpoint(KafkaMessageChannelBinder.java:163) ~[spring-cloud-stream-binder-kafka-3.2.4.jar:3.2.4]
	at org.springframework.cloud.stream.binder.AbstractMessageChannelBinder.doBindConsumer(AbstractMessageChannelBinder.java:426) ~[spring-cloud-stream-3.2.4.jar:3.2.4]
	... 24 common frames omitted
```

Attached is a project that reproduces the error.
[demoStreamKafka.zip](https://github.com/spring-cloud/spring-cloud-stream/files/8807176/demoStreamKafka.zip)

When configuring Spring Cloud version `2021.0.2` the project starts correctly.

**Comment by ferblaca** [comment]
> @ferblaca Thanks for reporting this issue (and the sample app). Turned out to be a bug that was introduced recently. See [this PR](https://github.com/spring-cloud/spring-cloud-stream/pull/2418) that fixes the issue. We will backport the changes to `3.2.x`.

Hi @sobychacko,

I have updated the spring-cloud version from `2021.0.3` (spring-cloud-stream-binder-kafka-core:3.2.4) to `2021.0.4` (spring-cloud-stream-binder-kafka-core:3.2.5) but this issue is still not resolved... You can check it with the demo application I attached to the issue and by modifying the spring-cloud version.

Can you check it?

**Comment by sobychacko** [comment]
@ferblaca On a quick look, we missed backporting this change to `3.2.x`, which means that the change only went into `4.0.x`. We are planning another `3.2.x` release soon and we can include this there. cc @olegz

**Comment by sobychacko** [comment]
@ferblaca This is now backported to `3.2.x`. Could you try the fix on the latest snapshot version (`3.2.6-SNAPSHOT`) of the binder?

**Comment by sobychacko** [comment]
Thank you for confirming. We will release it around end of October.
